Add support for spec compatible paths in generated server code

ofpiyush opened the pull request:
*Issue #, if available:*
#244 
*Description of changes:*
* Adds support for spec compatible paths in generated server code.
* Updates protoc to 3.12.4

No backwards incompatible changes have been made. Specifically:
* Go generated clients haven't been changed and will continue to send requests to the same URL as before.
* Go generated server will continue to serve on the CamelCased URLs it did before.
* `<ServiceName>PathPrefix` constant is unchanged as of now.
* `svc.PathPrefix()` still returns the CamelCased value.


Caveat: If someone is using a mux, they'll need to add the literal cased URL for handling manually as the intent was to not add anything to the API surface. [For context: start reading from here](https://github.com/twitchtv/twirp/pull/257#issuecomment-675220875)



By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

dpolansky commented on protoc-gen-twirp/generator.go:
```diff
@@ -1067,6 +1067,13 @@ func pathPrefix(file *descriptor.FileDescriptorProto, service *descriptor.Servic
+
+// pathForCamelCompat returns the old CamelCase complete path for requests to a
+// particular method on a particular service.
+// This method is here to ensure compatibility with older clients.
```
This comment isn't quite accurate, since clients will still make requests with camel-cased paths. This ultimately should link to documentation around this issue (which doesn't exist just yet).

I need to poke some folks internally about the proper place for these docs, and then we can update this comment to direct users to it.

dpolansky replied:
I made #262, which will add a page to the docs about this known issue.

dpolansky commented on protoc-gen-twirp/generator.go:
```diff
@@ -1104,8 +1111,13 @@ func (t *twirp) generateServerRouting(servStruct string, file *descriptor.FileDe
+		pathCamel := pathForCamelCompat(file, service, method)
 		methName := "serve" + stringutils.CamelCase(method.GetName())
-		t.P(`  case `, strconv.Quote(path), `:`)
+		caseCondition := strconv.Quote(path)
```
Could you add a test in `internal/twirptest` that verifies this new behavior? You will also need to run `make` to re-generate the code with these changes

ofpiyush commented:
`protobuf@3.11.0` is not available on brew anymore.

`protobuf@3.12.4` didn't seem to break anything so I've updated that in this PR as well. 

stdin stdout code in `pycompat` broke with python 3.8 on my local so I've updated it.

Let me know if I should force install 3.11.0 or change pycompat in any way.

dpolansky commented on internal/twirptest/snake_case_names/compatibility_test.go:
```diff
@@ -0,0 +1,68 @@
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package snake_case_names
```
I think we should just create a separate test for this instead of reusing this existing test scenario for clarity. In addition, it would allow us to test the literal names for both the service and method names. As it stands, `TestCompatibilityWithSpecCompatibleClient` only tests the literal method name and not the service name.

ofpiyush replied:
I assume this comment means a green signal for adding the new method `RegisterHaberdasherServer`.

Awesome!

dpolansky replied:
What is this referring to? I don't know what `RegisterHaberdasherServer` means.

ofpiyush replied:
First half of this comment. https://github.com/twitchtv/twirp/issues/244#issuecomment-666046457

While you're at it, please also go through the first half of this comment as it can completely resolve the issue.

https://github.com/twitchtv/twirp/issues/244#issuecomment-673935091

ofpiyush replied:
Resolved this to move conversation to main thread.

ofpiyush commented:
I've moved test to it's own package. Waiting for word on changes for service name and client hook.

I'll copy paste relevant parts from the comments to have context here.

## Server

This will need work from the user as well in cases where they're using a mux. We can manage method names without any changes from the user, but service name will need changes from the user.

One possible path which will contain this to 1 user change only, is to add a new method:

```go
example.RegisterHaberdasherServer(mux, handler)
```

Then we can keep changing code inside that function over releases and eventually into the new major version(if that ever happens) while making sure backwards compatibility is maintained.

## Client

The next best option I have is to generate a compatibility client hook and use a map to override URL path like I did on the test [here](https://github.com/ofpiyush/twirp/blob/312779fea1d5e901dd7bf906032eef1c78b8a570/internal/twirptest/snake_case_names/compatibility_test.go#L52-L57).

User experience could look something like:
```go
client := example.NewHaberdasherProtobufClient(url, client, twirp.WithClientHooks(example.CompatibilityHook))
```

dpolansky commented:
Thanks for the refresher @ofpiyush! I read those comments, but didn't quite understand what you meant with regard to the server-side aspect until now. I think we should proceed with just addressing the method names for now, as addressing the service names will be complex and expands the API surface-area, which we try to be cautious about. Does that sound reasonable to you?

With regard to the compatibility hook, I would be hesitant to recommend it as a general solution to this problem because although it gets the job done, the hooks aren't intended for this purpose and it could lead to confusion. Hooks are generally not intended to mutate requests or responses that significantly. For now, I'd feel better about telling users to change their protobuf definitions if this issue affects them than tell them to use hooks to solve the problem. My opinion is that if this issue is painful enough for many Twirp users, we would then release a new major version with either changes to the spec or changes to the Go implementation.

---

Thanks for moving the test case to a separate package. Once we publish the docs PR, could we just update that comment to direct users there as we discussed previously? Thanks again for addressing the feedback.

ofpiyush commented:
> I think we should proceed with just addressing the method names for now

We could do slightly better and add code to handle it in non-mux cases and we could drop a line of gotcha of documentation for people who'd use a mux.

The next best option to a new method is to add a new PathPrefix constant.

Mux user experience would look something like:
```go
mux.Handle(example.HaberdasherPathPrefix, twirpHandler)
mux.Handle(example.HaberdasherSomeNamingConventionPathPrefix, twirpHandler)
```

I see the `RegisterHaberDasherServer` as a better alternative because that is a one time change for the user. It would open up path for other changes as well like #263 and eventually a smoother transition to v6.

> because although it gets the job done, the hooks aren't intended for this purpose and it could lead to confusion

Then I should change my test as well. What alternative would you suggest for the test?

The next best option I have is to do it as a http client wrapper. User experience would look something like:

```go
client := example.NewHaberdasherProtobufClient(url, example.HaberdasherCompatibilityClient(client))
```
---------------------------------------------------------------------------------
>  For now, I'd feel better about telling users to change their protobuf definitions

I think the fundamental assumption you're working with is that these poorly named services (even in go itself) are not in production or given out to third party.

I come from a world where some APIs are already public, our python servers were live with other python clients before go clients were introduced. We weren't even explicitly following protobuf style guide. We just happened to be fortunate that the random convention I followed isn't incompatible.

I think making that requirement part of the spec would be a much better alternative if adding small amounts of code to the API surface is not acceptable.

dpolansky commented:
@ofpiyush I'd like to focus on just making this incremental improvement to handle the method names for now.

> Then I should change my test as well. What alternative would you suggest for the test?
> 
> The next best option I have is to do it as a http client wrapper. User experience would look something like:
> 
> ```go
> client := example.NewHaberdasherProtobufClient(url, example.HaberdasherCompatibilityClient(client))
> ```

I think this is preferable to modifying the request through hooks, so the test could use this method. I'd like to avoid adding additional scope to this PR beyond that though, such as providing the HTTP client implementation in the generated code as you demonstrated here.

ofpiyush commented on internal/twirptest/spec_compatibility/compatibility_test.go:
```diff
@@ -0,0 +1,67 @@
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package spec_compatibility
```
That was where it was! Check the comments above, haha!

@dpolansky asked me to make it into another package.

marioizquierdo replied:
Oh, sorry for the confusion. Yes I read that comment but I think it was based on previous context (the PR had a larger scope before). If it is still not a good idea (or I didn't understand the comment well), then just rename this test. No big deal :P

ofpiyush replied:
The general vibe I got was all the future tests that will be necessary to ensure spec compatibility will go into that package as well.

ofpiyush replied:
I've moved it back.

marioizquierdo replied:
> The general vibe I got was all the future tests that will be necessary to ensure spec compatibility will go into that package as well.

I see. But I would argue that is not a good way to group tests (this may be just my opinion). You may end up having only one specific test (non-camel-cased-names) under a generic umbrella term (spec-compat). Or you may end up having too many unrelated tests under the same umbrella term. We just don't know, so it is usually best not to try guessing the future too much, and just name things for what they are. If things change too much in the future, well, then we should refactor, but it is better to refactor later, at the right time, when you know what works better. A good rule of thumb to know when it's time to refactor is the [rule of three](https://blog.codinghorror.com/rule-of-three/) 🤓

marioizquierdo commented on protoc-gen-twirp/generator.go:
```diff
@@ -1060,14 +1060,33 @@ func (t *twirp) generateServer(file *descriptor.FileDescriptorProto, service *de
+
+func pathForGoClient(file *descriptor.FileDescriptorProto, service *descriptor.ServiceDescriptorProto, method *descriptor.MethodDescriptorProto) string {
+	return pathPrefix(file, stringutils.CamelCase(service.GetName())) + stringutils.CamelCase(method.GetName())
+}
```
^ forgot to remove this one

ofpiyush replied:
aah yes! removed it now

marioizquierdo commented on protoc-gen-twirp/generator.go:
```diff
@@ -1078,7 +1097,7 @@ func (t *twirp) generateServerRouting(servStruct string, file *descriptor.FileDe
 	t.P(`// Requests are always: POST `, pathPrefixConst, `/method`)
 	t.P(`// It can be used in an HTTP mux to route twirp requests along with non-twirp requests on other routes.`)
-	t.P(`const `, pathPrefixConst, ` = `, strconv.Quote(pathPrefix(file, service)))
+	t.P(`const `, pathPrefixConst, ` = `, strconv.Quote(pathPrefix(file, stringutils.CamelCase(service.GetName()))))
```
This is the same as using `servName`, but it's fine, being explicit may be more readable.

ofpiyush replied:
Yes I've been trying to keep strings and go variables separate so that later one can change these easier if necessary.

marioizquierdo commented:
This PR will be part of the upcoming v7 release.
@ofpiyush are you happy with all the changes? I can merge this in the release branch and get it ready with all other changes.

ofpiyush commented:
Yes. Feel free to merge it.


V7?

marioizquierdo commented:
Yes, we are going to skip v6 because there was a previous [version draft](https://twitchtv.github.io/twirp/docs/spec_v6.html) that was never fully implemented.

ofpiyush commented:
This PR can safely be added to V5. Can we add client support in V7 as well?

marioizquierdo commented:
Yes, this is backwards compatible. And the client support needs to be backwards compatible as well (through a client option, with default value that works like before). I'm including this on next release for simplicity (I'll spend some time this week managing the release). If you want, please submit another PR with the client side of this so we can bundle it together with the v7 release 🙏  😺 . If you don't have time right now, it's fine, we can handle that part later as well.
